### PR TITLE
Add basic released event to slider

### PR DESCRIPTION
**Pull request by flukejones**
Not sur if this was the best way to go about it, but I needed the slider to have a released event available to prevent such things as multiple queued changes from blocking, or spawning many multiples of threads.

With this I can have a callback be called only when the mouse is released. I guess there are things like key events that will be missed by this however.

### CODE REVIEW

**Comment by tronical** [comment]
Unfortunately @FloVanGH is on leave right now. API wise, my feeling is that it might be easier to have a boolean property to configure Slider's behavior whether to invoke `changed()` live while dragging the mouse or only when released. @ogoffart what do you think?

**Comment by ogoffart** [comment]
> API wise, my feeling is that it might be easier to have a boolean property to configure Slider's behavior whether to invoke changed() live while dragging the mouse or only when released.

I still think we might want two callbacks. See this use-case:
```
Slider {
   changed => { update-preview(); }
   released => { perform-change(); }
}
```

**Comment by ogoffart** [comment]
> Fair point, but is released the right name? Shouldn't this also be invoked when the slider knob is moved by the keyboard?

Yeah, this is tricky. Other name such a `interaction-ended` or `value-finalized` aren't that great. But it is also going to be send on keyboard event.
(Maybe it should be sent on key released only so that long press of the arrow can change the value, but only release emit the released callback)

**Comment by flukejones** [comment]
>  (Maybe it should be sent on key released only so that long press of the arrow can change the value, but only release emit the released callback)

This is what i think most people would expect (and what I would prefer).

As for the name perhaps `slider-released`? Personally I would like the name to convey that it is an event caused only by key/mouse-button event `up`. Many things like SDL use `the state of the key; SDL_PRESSED or SDL_RELEASED`

**Comment by tronical** [comment]
Flutter has two callbacks for this, `onChangeStart` and `onChangeEnd`. That's not really a pattern we use, but in the light of that a plain `released` callback that's invoked when the key is released as well as the mouse (or finger lifted) makes sense to me.

**Comment by flukejones** [comment]
@tronical I'll try to do some but no promises. As always I'm under pressure at work, buit since I need this for asusctl also I'll try to find the time.

**Comment by tronical** [comment]
No worries. Just let me know when you'd like me to take over (at any point).

**Comment by tronical** [comment]
I think that this is ready for review. But since I had my dirty hands in it, it would be good if @ogoffart could do the honours of finding bugs in the patches I added on top ;-)

File: tests/cases/widgets/slider_basic.slint [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+/*
+
+
+```rust
+use slint::platform::Key;
+use std::cell::RefCell;
+use std::rc::Rc;
+
+let instance = TestCase::new().unwrap();
+slint_testing::send_mouse_click(&instance, 5., 5.);
```
Comment (ogoffart) [inline tests/cases/widgets/slider_basic.slint:29]: should that mouse click be sent after we set the on_released callback so we check that the callback is emit (or not) when the mouse is released.

(would be nice to also test the changed signal at the same time)

File: internal/compiler/widgets/common/slider-base.slint [hunk 2]
```diff
@@ -84,6 +89,14 @@ export component SliderBase {
 
             reject
         }
+
+        key-released => {
+            if (!self.enabled) {
+                return reject;
+            }
+            root.released(root.value);
```
Comment (ogoffart) [inline internal/compiler/widgets/common/slider-base.slint:97]: should we not check that this is an arrow?

**Comment by flukejones** [comment]
I recovered the ref locally and pushed to https://github.com/flukejones/sixtyfps/tree/feat-slider-release-event

**Comment by tronical** [comment]
Continued in https://github.com/slint-ui/slint/pull/4821

File: internal/compiler/widgets/common/slider-base.slint (hunk 2 above)
Comment (tronical) [reply internal/compiler/widgets/common/slider-base.slint]: Good catch. Covered and tested.

**Pull request closed without merging**